File: 2001-3000/2817-minimum-absolute-difference-between-elements-with-constraint/2817-minimum-absolute-difference-between-elements-with-constraint.java
class Solution {
    public int minAbsoluteDifference(List<Integer> nums, int x) {
        TreeSet<Integer> set = new TreeSet<>((o1, o2) -> o1 - o2);
        int res = Integer.MAX_VALUE;
        for(int i = x; i < nums.size(); i++) {
            set.add(nums.get(i - x));
            int curr = nums.get(i);
            Integer low = set.floor(curr);
            Integer high = set.ceiling(curr);
            if (low != null) {
                res = Math.min(res, Math.abs(low - curr));
            }
            if (high != null) {
                res = Math.min(res, Math.abs(high - curr));
            }
        }
        return res;
    }
}

File: 2001-3000/2817-minimum-absolute-difference-between-elements-with-constraint/2817-minimum-absolute-difference-between-elements-with-constraint.cpp
class Solution {
public:
    int minAbsoluteDifference(vector<int>& nums, int x) {
        int n = nums.size();

        if (x == 0)
            return 0;

        set<int> st;
        int ans = INT_MAX;

        for (int i = n - 1 - x; i >= 0; i--) {
            st.insert(nums[i + x]);

            auto it = st.lower_bound(nums[i]);

            if (it != st.end())
                ans = min(ans, abs(nums[i] - *it));

            if (it != st.begin()) {
                --it;
                ans = min(ans, abs(nums[i] - *it));
            }
        }

        return ans;
    }
};